python/common_tool_methods.py

```diff
@@ -0,0 +1,146 @@
+    return f"{version}{fancy_suffix}"
+
+
+def get_version_number_with_project_name(project_name: str, version: str, fancy: bool, fancy_release_count: int) -> str:
+    fancy_suffix = f"-{fancy_release_count}" if fancy else ""
+    return f"{version}.{project_name}{fancy_suffix}"
+
+
+def get_template_environment(template_dir: str) -> Environment:
+    file_loader = FileSystemLoader(template_dir)
+    env = Environment(loader=file_loader)
+    return env
+
+
+def find_nth_overlapping(subject_string, search_string, n) -> int:
+    start = subject_string.find(search_string)
+
+    while start >= 0 and n > 1:
+        start = subject_string.find(search_string, start + 1)
+        n -= 1
+    return start
+
+
+def find_nth_overlapping_line_by_regex(subject_string, regex_pattern, n) -> int:
+    lines = subject_string.splitlines()
+    counter = 0
+    for i in range(len(lines)):
+        if re.match(regex_pattern, lines[i]):
+            counter = counter + 1
+        if counter == n:
+            return i
+    return -1
+
+
+def remove_string_inside_parentheses(param: str) -> str:
+    return re.sub(r"[(\[].*?[)\]]", "", param)
+
+
+def run(command, *args, **kwargs):
+    result = subprocess.run(command, *args, check=True, shell=True, **kwargs)
+    return result
+
+
+def cherry_pick_prs(prs: List[PullRequest.PullRequest]):
+    for pr in prs:
+        commits = pr.get_commits()
+        for single_commit in commits:
+            cp_result = run(f"git cherry-pick {single_commit.commit.sha}")
+            print(f"Cherry pick result for PR No {pr.number} and commit sha {single_commit.commit.sha}: {cp_result}  ")
+
+
+def get_version_details(version: is_version(str)) -> Dict[str, str]:
+    version_parts = version.split(".")
+    return {"major": version_parts[0], "minor": version_parts[1], "patch": version_parts[2]}
+
+
+def get_default_upcoming_version(version: is_version(str)) -> str:
+    project_version_details = get_version_details(version)
+    return f'{project_version_details["major"]}.{project_version_details["minor"]}.' \
+           f'{str(int(project_version_details["patch"]) + 1)}'
+
+
+def get_upcoming_minor_version(version: is_version(str)) -> str:
+    upcoming_version_details = get_version_details(version)
+    return f'{upcoming_version_details["major"]}.{upcoming_version_details["minor"]}'
+
+
+def is_major_release(version: is_version(str)) -> bool:
+    version_info = get_version_details(version)
+    return version_info["patch"] == "0"
+
+
+def str_array_to_str(str_array: List[str]) -> str:
+    return f"{os.linesep.join(str_array)}{os.linesep}"
+
+
+def get_prs(repo: Repository.Repository, earliest_date: datetime, base_branch: str, last_date: datetime = None):
+    pulls = repo.get_pulls(state="all", base=base_branch, sort="created", direction="desc")
+    picked_pulls = []
+    for pull in pulls:
+        if not pull.is_merged():
+            continue
+        if pull.created_at < earliest_date:
+            break
```

Comment:
I changed the logic . I added a filter logic instead of while loop. Now it is leaner

Reply:
1) seems that we still compare created_at with something.
I don't think we need to check creation date, but the merge date.

2) We will anyway check if pr is merged, but I think we can use `state='closed'` instead of 'all'

3) List comprehension is a nice python feature but I think having that much if's in an in-line expression decreases the readability.

4) Looking into our previous perl script, passing `sort='created'` doesn't seem to be enough.
I think we need to sort the pr's by their merge date. On the other hand, github api doesn't seem to be providing that

In the end, I would imagine this function as such:
```python
def get_prs_for_patch_release(repo: Repository.Repository, earliest_date: datetime, base_branch: str, last_date: datetime = None):
    pull_requests = repo.get_pulls(state="closed", base=base_branch, sort="created", direction="desc")

    # filter pull requests according to given time interval
    filtered_pull_requests = list()
    for pull_request in pull_requests:
        if not pull_request.is_merged():
            continue
        if p.merged_at < earliest_date:
            continue
        if last_date and p.merged_at > last_date:
            continue

        filtered_pull_requests.append(pull_request)

    # finally, sort the pr's by their merge date
    sorted_pull_requests = sorted(filtered_pull_requests,
                                  key=lambda pull_request: pull_request.merged_at)
    return sorted_pull_requests
```

Reply:
Fixed and willl be tested

Reply:
Fixed